rz-svd/rz-svd.c

```diff
@@ -0,0 +1,309 @@
+
+	cur = value;
+	value[0] = 0;
+
+	if (!ta->baseaddress[0]) {
+		while (!ta_iter_done(ta) &&
+			(yxml_parse(&ta->x, *ta->ptr) != YXML_ELEMSTART ||
+				strcasecmp(ta->x.elem, "peripherals"))) {
+			ta->ptr++;
+		}
+		if (ta_iter_done(ta))
+			return NULL;
+		ta_start = ta->ptr;
+		ta_x = ta->x;
+
+		level = 0;
+		while (!ta_iter_done(ta) && !ta->baseaddress[0]) {
+			switch ((r = yxml_parse(&ta->x, *ta->ptr))) {
+			case YXML_ELEMSTART:
+				level += 1;
+				if (level == 2 && strcasecmp(ta->x.elem, "baseAddress") == 0) {
+					cur = value;
+					*cur = 0;
+				}
+				break;
+
+			case YXML_ELEMEND:
+				level -= 1;
+				if (level < 0) {
+					return ta_iter_next(ta);
+				} else if (level == 2 && cur) {
+					strcpytrunc(ta->baseaddress, value, sizeof(ta->baseaddress));
+					cur = NULL;
+
+					// go back to the beginning of peripherals?
+					ta->ptr = ta_start;
+					ta->x = ta_x;
+				}
+				break;
+
+			case YXML_CONTENT:
+				if (!cur || level != 2) {
+					break;
+				}
+				tmp = ta->x.data;
+				while (*tmp && cur < value + sizeof(value)) {
+					*cur++ = *tmp++;
+				}
+				if (cur >= value + sizeof(value)) {
+					cur = NULL;
+				} else {
+					*cur = 0;
+				}
+				break;
+			default:
+				break;
+			}
+			ta->ptr++;
+		}
+		if (ta_iter_done(ta))
+			return NULL;
+	}
+	assert(ta->baseaddress[0]);
+
+	// find <register>
+	level = 0;
+	while (!ta_iter_done(ta)) {
+		r = yxml_parse(&ta->x, *ta->ptr);
+
+		if (r == YXML_ELEMSTART) {
+			level += 1;
+			if (level == 3 &&
+				strcasecmp(ta->x.elem, "register") == 0) {
+				break;
+			}
+
+		} else if (r == YXML_ELEMEND) {
+			level -= 1;
+			if (level < 1) {
+				ta->baseaddress[0] = 0;
+				return ta_iter_next(ta);
+			}
+		}
+		ta->ptr++;
+	}
+	if (ta_iter_done(ta)) {
+		return NULL;
+	}
+
+	assert(r == YXML_ELEMSTART);
+
+	cur = NULL;
+	value[0] = 0;
+	elem_type = -1;
+
+	*ta->bitoffset = *ta->regname = *ta->bitwidth = 0;
+
+	for (;;) {
+		switch (r) {
+		case YXML_ELEMSTART:
+			level += 1;
+			if (level != 5)
+				break;
+			if (strcasecmp(ta->x.elem, "displayName") == 0) {
+				elem_type = REGNAME;
+			} else if (strcasecmp(ta->x.elem, "description") == 0) {
+				elem_type = DESCRIPTION;
+			} else if (strcasecmp(ta->x.elem, "addressOffset") == 0) {
+				elem_type = BITOFFSET;
+			} else if (strcasecmp(ta->x.elem, "size") == 0) {
+				elem_type = BITWIDTH;
+			} else {
+				break;
+			}
+			cur = value;
+			*cur = 0;
+			break;
+
+		case YXML_ELEMEND:
+			level -= 1;
+			if (level < 2) {
+				ta->baseaddress[0] = 0;
+				return ta_iter_next(ta);
+			} else if (level != 4 || !cur) {
+				break;
+			}
+			cur = NULL;
+			switch (elem_type) {
+			case BITOFFSET:
+				strcpytrunc(ta->bitoffset, value, sizeof(ta->bitoffset));
+				break;
+			case BITWIDTH:
+				strcpytrunc(ta->bitwidth, value, sizeof(ta->bitwidth));
+				break;
+			case REGNAME:
+				strcpytrunc(ta->regname, value, sizeof(ta->regname));
+				break;
+			case DESCRIPTION:
+				rz_str_replace_char(value, '\n', ' ');
```

Comment:
I guess this is also relevant to https://github.com/rizinorg/rizin/issues/1356

Reply:
I don't think so, this is just to handle descriptions like
```xml
              <description>Port x configuration bits (y =
              0..15)</description>
```